Srcs/heiftojpeg/heiftojpeg.cpp

```diff
@@ -0,0 +1,183 @@
+            // cout << reader.getPropertyLhvc(contextId, property.index);
+        } else if (property.type == ImageFileReaderInterface::ItemPropertyType::LSEL) {
+            cout << "LSEL ";
+            cout << reader.getPropertyLsel(contextId, property.index).layerId;
+        } else if (property.type == ImageFileReaderInterface::ItemPropertyType::OINF) {
+            cout << "OINF ";
+            // cout << reader.getPropertyOinf(contextId, property.index);
+        } else if (property.type == ImageFileReaderInterface::ItemPropertyType::RLOC) {
+            cout << "RLOC ";
+            cout << reader.getPropertyRloc(contextId, property.index).horizontalOffset;
+        } else if (property.type == ImageFileReaderInterface::ItemPropertyType::TOLS) {
+            cout << "TOLS ";
+            cout << reader.getPropertyTols(contextId, property.index).targetOlsIndex;
+        } else {
+            cout << "unknown ";
+        }
+        cout << "\n";
+    }
+
+    // Always reuse the parameter set from the first tile, sometimes tile 7 or 8 is corrupted
+    HevcImageFileReader::ParameterSetMap parameterSet;
+    reader.getDecoderParameterSets(contextId, tileItemIds.at(0), parameterSet);
+    std::string codeType = reader.getDecoderCodeType(contextId, tileItemIds.at(0));
+    ImageFileReaderInterface::DataVector parametersData;
+    if ((codeType == "hvc1") || (codeType == "lhv1")) {
+        // VPS (HEVC specific)
+        parametersData.insert(parametersData.end(), parameterSet.at("VPS").begin(), parameterSet.at("VPS").end());
+    }
+
+    if ((codeType == "avc1") || (codeType == "hvc1") || (codeType == "lhv1")) {
+        // SPS and PPS
+        parametersData.insert(parametersData.end(), parameterSet.at("SPS").begin(), parameterSet.at("SPS").end());
+        parametersData.insert(parametersData.end(), parameterSet.at("PPS").begin(), parameterSet.at("PPS").end());
+    } else {
+        // No other code types supported
+        // throw FileReaderException(FileReaderException::StatusCode::UNSUPPORTED_CODE_TYPE);
+    }
+
+    ImageFileReaderInterface::DataVector itemDataWithDecoderParameters;
+    ImageFileReaderInterface::DataVector itemData;
+
+    std::vector<Magick::Image> tileImages;
+    for (auto& tileItemId: tileItemIds) {
+        itemDataWithDecoderParameters.clear();
+        itemDataWithDecoderParameters.insert(itemDataWithDecoderParameters.end(), parametersData.begin(), parametersData.end());
+
+        itemData.clear();
+        reader.getItemData(contextId, tileItemId, itemData);
+
+        // +1 comes from skipping first zero after decoder parameters
+        itemDataWithDecoderParameters.insert(itemDataWithDecoderParameters.end(), itemData.begin() + 1, itemData.end());
+
+        Magick::Image image;
+        decodeData(itemDataWithDecoderParameters, &image);
+        tileImages.push_back(image);
+    }
+
+    Magick::Montage montageOptions;
+    montageOptions.tile("8x6");
+    montageOptions.geometry("512x512");
```

Comment:
these should in theory be computable from the metadata, but I can't get the geometry info to come out correctly.

Reply:
another fork does indeed find the number of tiles, rows, and columns from the metadata (gridItem.columnsMinusOne, etc.) but also hard-codes the geometry 
https://github.com/nokiatech/heif/compare/master...Eriken:master#diff-3ef7695ca549c96689462fca2ffeaaa3R10 

it would probably be simpler to execute the imagemagick montage binary like them instead of bothering to link to the ImageMagick library given that we're doing that with ffmpeg and writing the tiles to disk anyways, but it's fine as is since you already did the cmake config

Reply:
i also pull out columnsMinusOne (look above in a log line) but it's always empty. I'm not sure if the reader is buggy or if apple has implemented the grid item differently than the standard.

I prefer the direct C++ bindings as we benefit from static type checking with the inputs. Also I don't need to have all 48 tiles on disk at the same time, they are all pulled into memory one by one as ffmpeg finishes generating them.